### PR TITLE
SDIT-902 Add incentive levels and allocations to the test data DSL

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderBookingBuilder.kt [hunk 1]
```diff
@@ -79,4 +79,15 @@ class OffenderBookingBuilder(
...
+  override fun incentive(
+    iepLevelCode: String,
+    userId: String?,
+    sequence: Long,
+    commentText: String,
+    auditModuleName: String?,
+    iepDateTime: LocalDateTime,
+  ) {
+    this.incentives += IncentiveBuilder(iepLevelCode, userId, sequence, commentText, auditModuleName, iepDateTime)
+  }
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderBookingBuilder.kt:92]: Main change 1: include incentives in the offender booking DSL

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/DSL.kt [hunk 2]
```diff
@@ -68,6 +69,31 @@ class TestData(private val repository: Repository) : TestDataDsl {
...
+    offenderBooking: OffenderBooking,
+    courseActivity: CourseActivity,
+    startDate: String?,
+    programStatusCode: String,
+    endDate: String?,
+    payBands: List<OffenderProgramProfilePayBandBuilder>,
+    endReasonCode: String?,
+    endComment: String?,
+    dsl: CourseAllocationDsl.() -> Unit,
+  ): OffenderProgramProfile = repository.save(
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/DSL.kt:84]: Main change 1: add course allocations to the top level of the test data DSL

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/DSL.kt [hunk 3]
```diff
@@ -135,8 +174,21 @@ interface BookingDsl {
...
+
+  @IncentiveDslMarker
+  fun incentive(
+    iepLevelCode: String = "ENT",
+    userId: String? = null,
+    sequence: Long = 1,
+    commentText: String = "comment",
+    auditModuleName: String? = null,
+    iepDateTime: LocalDateTime = LocalDateTime.now(),
+  )
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/DSL.kt:186]: Main change 2: add incentives to the offender booking DSL

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt [hunk 4]
```diff
@@ -19,43 +19,30 @@ import org.springframework.http.MediaType
...
 import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.PayPerSession
 import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.SlotCategory
 import java.math.BigDecimal
 import java.time.LocalDate
 import java.time.LocalDateTime
 
-private const val PRISON_ID = "LEI"
-private const val ROOM_ID: Long = -8 // random location from R__3_2__AGENCY_INTERNAL_LOCATIONS.sql
-private const val PROGRAM_CODE = "INTTEST"
-private const val IEP_LEVEL = "STD"
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:44]: Decided these constants weren't helping much, particularly in tests with different values for those entities.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt [hunk 5]
```diff
@@ -19,43 +19,30 @@ import org.springframework.http.MediaType
...
 
   @Autowired
   private lateinit var repository: Repository
 
-  @Autowired
-  private lateinit var offenderProgramProfileBuilderFactory: OffenderProgramProfileBuilderFactory
-
-  @Autowired
-  private lateinit var offenderProgramProfilePayBandBuilderFactory: OffenderProgramProfilePayBandBuilderFactory
-
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:58]: All allocations are now created via the test data DSL

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AttendanceResourceIntTest.kt [hunk 6]
```diff
@@ -91,18 +86,20 @@ class AttendanceResourceIntTest : IntegrationTestBase() {
...
+        offenderBooking = offender.latestBooking()
         programService {
           courseActivity = courseActivity {
             payRate()
             courseSchedule()
             courseScheduleRule()
           }
+          courseSchedule = courseActivity.courseSchedules.first()
         }
+        allocation = courseAllocation(offenderBooking, courseActivity)
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AttendanceResourceIntTest.kt:101]: Most tests use an allocation so create on in the set up. Any tests needing a different or no allocation can set up from scratch.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderProgramProfileBuilder.kt [hunk 7]
```diff
@@ -1,51 +1,23 @@
 package uk.gov.justice.digital.hmpps.nomisprisonerapi.helper.builders
 
-import org.springframework.stereotype.Component
 import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.CourseActivity
 import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderBooking
 import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderProgramProfile
 import java.time.LocalDate
 
-@Component
-class OffenderProgramProfileBuilderFactory(
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderProgramProfileBuilder.kt:10]: All data creation is now done through the test data DSL so can get rid of the old builder factories

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt [hunk 8]
```diff
@@ -752,15 +744,17 @@ class ActivityResourceIntTest : IntegrationTestBase() {
...
-        val offender =
-          repository.save(OffenderBuilder(nomsId = "A1234TT").withBooking(OffenderBookingBuilder(agencyLocationId = PRISON_ID, incentives = listOf(IncentiveBuilder(iepLevel = "STD")))))
-        repository.save(
-          offenderProgramProfileBuilderFactory.builder(
-            payBands = listOf(offenderProgramProfilePayBandBuilderFactory.builder(endDate = LocalDate.now().minusDays(1).toString())),
-          ),
-          offender.latestBooking(),
-          courseActivity,
-        )
+        testData(repository) {
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:747]: Hmm - should we not write it like this ?

`
       testData(repository) {
          offender = offender(nomsId = "A1234TT") {
            booking(agencyLocationId = "LEI") {
              incentive(iepLevelCode = "STD")
              courseAllocation(courseActivity) {
                payBand(endDate = yesterday.toString())
              }
            }
          }
        }`

**Review by andymarke — APPROVED** [review]
LGTM

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt (hunk 8 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:747]: As per Slack conv. courseAllocation could live under either an offender booking or under a course activity - and as you rightly pointed out why not allow both? Coming soon.

**Pull request merged**